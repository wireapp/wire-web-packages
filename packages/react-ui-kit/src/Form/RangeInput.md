Demo:

```js
import {Columns, Column, RangeInput} from '@wireapp/react-ui-kit';
import {useState} from 'react';

const [zoom, setZoom] = useState(1);

const handleChange = e => {
  setZoom(+e.target.value);
};

<>
  <div>
    <RangeInput
      label={'Zoom'}
      minValueLabel="-"
      maxValueLabel="+"
      onChange={handleChange}
      value={zoom}
      min={1}
      max={3}
      step={0.1}
    />
  </div>
</>;
```
