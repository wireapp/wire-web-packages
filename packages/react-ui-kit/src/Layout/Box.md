Demo:

```js
import {Box} from '@wireapp/react-ui-kit';

<Box>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce a mattis nibh, sed maximus leo. Fusce a lacinia sem,
  vitae ornare dolor. Quisque rhoncus, magna non lacinia sagittis, erat augue fringilla metus, eu consectetur leo velit
  non lacus. Phasellus ipsum turpis, dapibus ut purus in, lobortis consectetur mi.
</Box>;
```
