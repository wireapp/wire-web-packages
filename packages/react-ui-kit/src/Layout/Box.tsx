/*
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

/** @jsx jsx */
import {ObjectInterpolation, jsx} from '@emotion/core';
import React from 'react';
import {COLOR} from '../Identity';

export interface BoxProps<T = HTMLDivElement> extends React.HTMLProps<T> {}

const boxStyles: (props: BoxProps) => ObjectInterpolation<undefined> = props => ({
  border: `2px solid ${COLOR.GRAY_LIGHTEN_72}`,
  borderradius: '8px',
  padding: '16px 32px',
});

const Box = (props: BoxProps) => <div css={boxStyles(props)} {...props} />;

export {Box, boxStyles};
