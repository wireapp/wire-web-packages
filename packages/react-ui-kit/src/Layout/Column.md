Demo:

```js
import {Container, Columns, Column, COLOR} from '@wireapp/react-ui-kit';

const ColumnsStyle = {
  marginBottom: '12px',
};

const ColumnStyle = {
  backgroundColor: COLOR.GRAY_LIGHTEN_32,
  border: `1px solid ${COLOR.GRAY_LIGHTEN_24}`,
};

<Container>
  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
    <Column style={ColumnStyle}>Column</Column>
  </Columns>

  <Columns style={ColumnsStyle}>
    <Column style={ColumnStyle}>
      <Columns>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
      </Columns>
    </Column>
    <Column style={ColumnStyle}>
      <Columns>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
      </Columns>
    </Column>
    <Column style={ColumnStyle}>
      <Columns>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
        <Column style={ColumnStyle}>Column</Column>
      </Columns>
    </Column>
  </Columns>
</Container>;
```
