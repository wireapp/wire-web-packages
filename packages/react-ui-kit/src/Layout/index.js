import setGlobalStyles from '../globalStyles';

setGlobalStyles();

export * from './Column';
export * from './Container';
export * from './Content';
export * from './Footer';
export * from './Header';
export * from './StyledApp';
