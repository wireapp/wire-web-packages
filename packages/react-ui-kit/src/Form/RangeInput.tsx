/*
 * Wire
 * Copyright (C) 2022 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import {forwardRef} from 'react';

import {CSSObject} from '@emotion/react';

import {InputLabel} from './InputLabel';
import {
  getImageCropZoomInputStyles,
  getValueLabelStyles,
  rangeInputWrapperStyles,
  ValueLabelPosition,
} from './RangeInput.styles';

import {Theme} from '../Layout';
import {TextProps} from '../Text';

export interface RangeInputProps<T = HTMLInputElement> extends TextProps<T> {
  label?: string;
  minValueLabel?: string;
  maxValueLabel?: string;
  wrapperCSS?: CSSObject;
}

export const RangeInput = forwardRef<HTMLInputElement, RangeInputProps<HTMLInputElement>>(
  (
    {
      id = Math.random().toString(),
      label,
      minValueLabel,
      maxValueLabel,
      min = '0',
      max = '100',
      value = '0',
      onChange,
      wrapperCSS,
      ...inputProps
    },
    ref,
  ) => {
    const minNum = Number(min);
    const maxNum = Number(max);
    const valueNum = Number(value);

    const backgroundSize = `${((valueNum - minNum) * 100) / (maxNum - minNum)}% 100%` as const;

    return (
      <div css={wrapperCSS}>
        {label && <InputLabel htmlFor={id}>{label}</InputLabel>}
        <div css={rangeInputWrapperStyles}>
          {minValueLabel && (
            <span css={(theme: Theme) => getValueLabelStyles(theme, ValueLabelPosition.LEFT)}>{minValueLabel}</span>
          )}
          {maxValueLabel && (
            <span css={(theme: Theme) => getValueLabelStyles(theme, ValueLabelPosition.RIGHT)}>{maxValueLabel}</span>
          )}
          <input
            ref={ref}
            css={(theme: Theme) => getImageCropZoomInputStyles(theme, backgroundSize)}
            id={id}
            name={id}
            min={min}
            max={max}
            value={value}
            onChange={onChange}
            type="range"
            {...inputProps}
          />
        </div>
      </div>
    );
  },
);
RangeInput.displayName = 'RangeInput';
