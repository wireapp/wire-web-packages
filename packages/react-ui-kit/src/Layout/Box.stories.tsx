/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import {Meta, StoryObj} from '@storybook/react';

import {Box} from './Box';

import {COLOR} from '../Identity/colors';
import {Text} from '../Text';

const meta: Meta<typeof Box> = {
  component: Box,
  title: 'Layout/Box',
  decorators: [
    Story => (
      <div style={{padding: '24px', maxWidth: '600px', margin: '0 auto'}}>
        <Story />
      </div>
    ),
  ],
};

export default meta;
type Story = StoryObj<typeof Box>;

const LOREM_IPSUM = `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce a mattis nibh, sed maximus leo. Fusce a lacinia sem,
vitae ornare dolor. Quisque rhoncus, magna non lacinia sagittis, erat augue fringilla metus, eu consectetur leo velit
non lacus. Phasellus ipsum turpis, dapibus ut purus in, lobortis consectetur mi.`;

export const Default: Story = {
  args: {
    children: LOREM_IPSUM,
  },
};

export const WithCustomPadding: Story = {
  args: {
    children: LOREM_IPSUM,
    style: {padding: '32px'},
  },
};

export const WithCustomBorder: Story = {
  args: {
    children: LOREM_IPSUM,
    style: {border: `2px solid ${COLOR.BLUE}`},
  },
};

export const WithFormattedContent: Story = {
  args: {
    children: (
      <>
        <Text block style={{fontSize: '20px', fontWeight: 'bold', marginBottom: '16px'}}>
          Important Notice
        </Text>
        <Text block style={{marginBottom: '12px'}}>
          {LOREM_IPSUM}
        </Text>
        <Text block style={{color: COLOR.GRAY}}>
          Last updated: 2025-01-01
        </Text>
      </>
    ),
  },
};

export const Nested: Story = {
  args: {
    children: (
      <>
        <Text block style={{marginBottom: '16px'}}>
          Outer box content
        </Text>
        <Box style={{backgroundColor: COLOR.GRAY_LIGHTEN_92}}>
          <Text>Inner box content</Text>
        </Box>
      </>
    ),
  },
};

export const Interactive: Story = {
  args: {
    children: (
      <div style={{textAlign: 'center'}}>
        <Text block style={{marginBottom: '16px'}}>
          Click the button below
        </Text>
        <button
          style={{
            backgroundColor: COLOR.BLUE,
            border: 'none',
            borderRadius: '4px',
            color: COLOR.WHITE,
            cursor: 'pointer',
            padding: '8px 16px',
          }}
        >
          Action Button
        </button>
      </div>
    ),
    style: {cursor: 'pointer'},
  },
};
