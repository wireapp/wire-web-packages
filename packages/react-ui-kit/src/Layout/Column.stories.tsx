/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import {ReactNode} from 'react';

import {Meta, StoryObj} from '@storybook/react';

import {Column, Columns} from './Column';

import {COLOR} from '../Identity/colors';
import {QueryKeys} from '../mediaQueries';

const meta: Meta<typeof Columns> = {
  component: Columns,
  title: 'Layout/Column',
};

export default meta;
type Story = StoryObj<typeof Columns>;

const Box = ({children}: {children: ReactNode}) => (
  <div
    style={{
      backgroundColor: COLOR.GRAY_LIGHTEN_72,
      border: `1px solid ${COLOR.GRAY_LIGHTEN_48}`,
      borderRadius: '4px',
      padding: '16px',
      textAlign: 'center',
    }}
  >
    {children}
  </div>
);

export const Default: Story = {
  args: {
    children: (
      <>
        <Column>
          <Box>Column 1</Box>
        </Column>
        <Column>
          <Box>Column 2</Box>
        </Column>
        <Column>
          <Box>Column 3</Box>
        </Column>
      </>
    ),
  },
};

export const TwoColumns: Story = {
  args: {
    children: (
      <>
        <Column>
          <Box>Column 1</Box>
        </Column>
        <Column>
          <Box>Column 2</Box>
        </Column>
      </>
    ),
  },
};

export const ResponsiveColumns: Story = {
  args: {
    query: QueryKeys.TABLET,
    children: (
      <>
        <Column>
          <Box>Stacks on tablet</Box>
        </Column>
        <Column>
          <Box>Stacks on tablet</Box>
        </Column>
        <Column>
          <Box>Stacks on tablet</Box>
        </Column>
      </>
    ),
  },
};

export const NestedColumns: Story = {
  args: {
    children: (
      <>
        <Column>
          <Box>Main Column</Box>
        </Column>
        <Column>
          <Columns>
            <Column>
              <Box>Nested 1</Box>
            </Column>
            <Column>
              <Box>Nested 2</Box>
            </Column>
          </Columns>
        </Column>
      </>
    ),
  },
};
