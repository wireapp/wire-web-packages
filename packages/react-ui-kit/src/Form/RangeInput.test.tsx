/*
 * Wire
 * Copyright (C) 2022 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 *
 */

import React from 'react';
import {matchComponent} from '../test/testUtil';
import {RangeInput, RangeInputProps} from './RangeInput';
import {fireEvent, render} from '@testing-library/react';
import {StyledApp, THEME_ID} from '../Layout';
import {matchers} from '@emotion/jest';
expect.extend(matchers);

const getDefaultProps = () => ({
  id: 'zoom-input',
  label: 'Zoom',
  minValueLabel: '-',
  maxValueLabel: '+',
  onChange: jest.fn(),
  value: 0,
  min: 1,
  max: 3,
  step: 0.1,
});

const ThemedRangeInput = (props: RangeInputProps) => (
  <StyledApp themeId={THEME_ID.LIGHT}>
    <RangeInput {...props} />
  </StyledApp>
);

describe('"RangeInput"', () => {
  it('matches snapshot', () => matchComponent(<ThemedRangeInput {...getDefaultProps()} />));

  it('renders label', () => {
    const props = getDefaultProps();
    const {getByLabelText} = render(<ThemedRangeInput {...props} />);
    expect(getByLabelText(props.label)).toBeDefined();
  });

  it('renders min and max value labels', () => {
    const props = getDefaultProps();
    const {getByText} = render(<ThemedRangeInput {...props} />);
    expect(getByText(props.minValueLabel)).toBeDefined();
    expect(getByText(props.maxValueLabel)).toBeDefined();
  });

  it('fires onChange callback', () => {
    const props = getDefaultProps();
    const {getByRole} = render(<ThemedRangeInput {...props} />);

    const rangeInput = getByRole('slider');
    fireEvent.change(rangeInput, {value: 2});

    expect(props.onChange).toHaveBeenCalled();
  });

  it("updates slider's background-size prop based on value", () => {
    const {getByRole, rerender} = render(<ThemedRangeInput {...getDefaultProps()} value={1} />);
    const rangeInput = getByRole('slider');

    expect(rangeInput).toHaveStyleRule('background-size', '0% 100%');

    rerender(<ThemedRangeInput {...getDefaultProps()} value={2} />);
    expect(rangeInput).toHaveStyleRule('background-size', '50% 100%');

    rerender(<ThemedRangeInput {...getDefaultProps()} value={3} />);
    expect(rangeInput).toHaveStyleRule('background-size', '100% 100%');
  });
});
