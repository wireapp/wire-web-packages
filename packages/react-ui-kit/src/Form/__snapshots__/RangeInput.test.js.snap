// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"RangeInput" matches snapshot 1`] = `
<div
  className="css-7uml9n-StyledAppContainer"
>
  <div
    className="css-7uml9n-StyledAppContainer"
  >
    <div
      className="css-0"
    >
      <label
        className="css-180k17q-InputLabel"
        htmlFor="zoom-input"
      >
        Zoom
      </label>
      <div
        className="css-irx2dz"
      >
        <span
          className="css-1xle7h4"
        >
          -
        </span>
        <span
          className="css-7g2ml5"
        >
          +
        </span>
        <input
          className="css-5xbpwv"
          id="zoom-input"
          max={3}
          min={1}
          name="zoom-input"
          onChange={[MockFunction]}
          step={0.1}
          type="range"
          value={0}
        />
      </div>
    </div>
  </div>
</div>
`;
